PURL-TYPES.rst

```diff
@@ -297,6 +297,22 @@ github
       pkg:github/package-url/purl-spec@244fd47e07d1004
       pkg:github/package-url/purl-spec@244fd47e07d1004#everybody/loves/dogs
 
+go
+------
+``go`` for Go modules:
+
+- The ``namespace`` field is empty and implies the go mod proxy.
+- The ``name`` will be the full module path.
+- The ``subpath`` will represent the package path within a module.
+- The ``version`` will be a valid go version or pseudoversion, or empty.
+- Additional Build information for binaries can be included as ``qualifiers`` (i.e VCS info, go version info, GoArch/GoOS info etc)
+- Examples::
+
+      pkg:go/google.golang.org%2Fgenproto#googleapis/api/annotations
+      pkg:go/github.com%2Fjmorion%2Fsqlx@v1.1.2#api
+      pkg:go/golang.org%2Fx%2Fvuln?goversion=1.23.2&vcs=git&vcs_modified=true#cmd/govulncheck
```

Comment:
There is a likely problem with the use of subpath: there is no way to determine where the module ends and the package starts in the general case, is there?
For instance, in the path `google.golang.org/genproto/googleapis/api/annotations` how can I determine safely that `google.golang.org/genproto` is a module and that `googleapis/api/annotations` is a package inside this module?  I need either a go proxy lookup or a full filesystem to locate a go.mod/go.sum file, right?

Reply:
There is a way, if the module's code is available to you, to determine from a package import where module path ends and the package path begins by making HTTP requests.

I think the use of the subpath here is good because it puts the burden of determining this on whatever generates the PURL, which is likely aware of Go and either has the module paths or is most likely to be able to find the module path from the full package path. Then if you want to use a tool that checks PURLs against a database of information about modules (eg vulnerabilities), the tool already has all the information it needs. Otherwise, either the tool would need to make external API calls to figure out the module path of the PURL or the database would need to have an entry for every package in the module.

Reply:
Just to add to @matt-phylum's comment. If a tool is producing a PURL for a Go artifact, then it can use `go version`, [Debug.BuildInfo](https://pkg.go.dev/debug/buildinfo#BuildInfo), or [packages.Load](https://pkg.go.dev/golang.org/x/tools/go/packages#Load) to get information about the package and its corresponding module. The encoding proposed here then makes it clear what the modules and packages are.

Reply:
I don't believe this is true for the general use case of PURLs. E.g. we do static analysis of binaries and while we can get information about linked packages, there's no indication of which part of the paths correspond to modules

Reply:
Right, if you are looking at a Go symbol from the symbol table, you can get its package. You can get the module correctly by prefix-matching it with module information from `debug.Buildinfo` of the binary, unless there are several modules that are prefixes of the package. My inclination is that it should not affect what is proposed here. (Arguably, there should be a way to get module information for a symbol in the binary, just the way one can do it for the source analysis.)